src/components/LibraryItem.tsx

```diff
@@ -73,31 +90,33 @@ export class LibraryItem extends React.Component<LibraryItemProps, LibraryItemSt
         let nestedElements = null;
         let clusteredElements = null;
 
-        if (this.state.expanded) // Show only nested elements when expanded.
-        {
-            if (this.props.data.childItems && (this.props.data.childItems.length > 0)) {
+        // visible only nested elements when expanded.
+        if (this.state.expanded && this.state.visible && this.props.data.childItems.length > 0) {
 
-                // Break item list down into sub-lists based on the type of each item.
-                let groupedItems = new GroupedItems(this.props.data.childItems);
+            // Break item list down into sub-lists based on the type of each item.
+            let groupedItems = new GroupedItems(this.props.data.childItems);
 
-                // There are some leaf nodes (e.g. methods).
-                clusteredElements = this.getClusteredElements(groupedItems);
+            // There are some leaf nodes (e.g. methods).
+            clusteredElements = this.getClusteredElements(groupedItems);
 
-                // There are intermediate child items.
-                nestedElements = this.getNestedElements(groupedItems);
-            }
+            // There are intermediate child items.
+            nestedElements = this.getNestedElements(groupedItems);
         }
 
-        return (
-            <div className={this.getLibraryItemContainerStyle()}>
-                <div className={"LibraryItemHeader"} onClick={this.onLibraryItemClicked.bind(this)} >
-                    {iconElement}
-                    <div className={libraryItemTextStyle}>{this.props.data.text}</div>
+        if (this.state.visible) {
+            return (
+                <div className={this.getLibraryItemContainerStyle()}>
+                    <div className={"LibraryItemHeader"} onClick={this.onLibraryItemClicked.bind(this)} >
+                        {iconElement}
+                        <div className={libraryItemTextStyle}>{this.props.data.text}</div>
+                    </div>
+                    {clusteredElements}
+                    {nestedElements}
                 </div>
-                {clusteredElements}
-                {nestedElements}
-            </div>
-        );
+            );
+        } else {
+            return null;
+        }
```

Comment:
Do we come all the way here (i.e. after computing all other stuff like `nestedElements`) just to `return null` when `this.state.visible === false`? Remember that will incur unnecessary calculations that can be avoided? If that's the case move `return null` all the way up (and remove other `if (this.state.visible)` related checks since it will be `true` once it passed the initial check.